### ISSUE TITLE
Callback URL - Authentication

### ISSUE THREAD

**Issue by Manu4771** [JUNGgroup/node-red-contrib-jung#3]
Hallo devs. 

Muss bei der Callback URL tatsächlich die locale Node-RED instalation aus dem freien Web verfügbar gemacht werden? 
Selbst als ich dies getan habe, verlangt Node-Red einen Login. Diesen kann ich aber nirgends in der Config pflegen. Was ist hier die Idee dahinter ? 

Fehlermeldeung: 
"registerOnJung> Registering `Thermostat_Wohnzimmer_Ist_temp.` (45c68ebc-af79-4c16-a828-0cc5cc04484f) for active feedback from JUNG on listener URL=[meine URL]"

**Comment by JUNGgroup** [comment]
> Muss bei der Callback URL tatsächlich die locale Node-RED instalation aus dem freien Web verfügbar gemacht werden?

Ja. Wenn etwas in der JUNG Open API Cloud passiert, muss die Cloud diese Benachrichtigung an Ihre lokale Installation weiterleiten.
Das bedeutet, dass die Callback-URL definiert und von der JUNG Open API Cloud aus zugänglich sein muss.


Zum Thema Login kann ich nichts sagen - normalerweise ist kein Login erforderlich.

**Comment by Manu4771** [comment]
Meine lokale Node-RED Installation ist aus dem Internet über eine feste URL erreichbar. Ich möchte das das Port-Forwarding in meinem Router vermeiden. Wenn ich diese URL aus dem freien Netz aufrufe muss ich mich authentifizieren (401). Ansonsten könnte ja jeder mit der URL auf mein Node-RED zugreifen. 

Ist die o.g. Fehlermeldung lediglich eine Bestätigung, dass es funktioert hat und kann der Datenaustausch auch ohne Autentifizierung erfolgen? 
Die Dokumentaion dazu ist leider sehr spärlich.

**Comment by patrickkaucher** [comment]
Hi @Manu4771 , konntest du das Problem lösen, ich hab das selbe Problem.

**Comment by Manu4771** [comment]
Nein, das Problem betseht weiterhin. Ich bekomme wie gesagt über die Debugs nur o.g. Meldung. Da ich ansonsten keine Payload angezeigt bekomme, gehe ich davon aus, dass das so nicht funktioniert. Mir fehlt also aktuell jeglicher Ansatz das Problem zu lösen. Neue Jung-Geräte bekomme ich aber z.b. sofort in Node-RED angezeigt, d.h. dieser Part über die API muss funktionieren.

**Comment by patrickkaucher** [comment]
Geht mir genau gleich.

**Comment by CrazyBird159** [comment]
Ich habe das gleiche Problem. Ich bekomme auch 401 Unauthorized.
Dies bekomme ich aber auch beim auf beim Aufruf aus dem lokalen Netz.

Für die Freigabe habe ich in der Fritzbox mit Portfreigabe (http://[myfritzid].myfritz.net:4711/jung/abcdefghijklmnopqrs | 4711 gemapped auf 1880) gearbeitet und mit der MyFritzfreigabe (http://[devicename].[myfritzid].myfritz.net:1880/jung/abcdefghijklmnopqrs).
Für beide Fälle komme ich dann über Hopscotch aus dem Internet auch die Unauthorized Meldung. Ich habe einfach mal einen random Bearer Token mitgeschickt, dann bekomme ich aber 500 Internal Server Error. Ich weiß hier nicht, ob beim anlegen der Subscription im Hintergund ein Token gespeichert wird, den die Jung Cloud zur Authentifizierung dann nutzt.

Auf https://open-api.prod.jung-hosting.de/Subscriptions sehe ich die Instanzen. In der Spalte Fehler, zählt der Counter für z.B. für Temperatur hoch. Über das "i" kann man sich Datum und Uhrzeit der Fehler anzeigen lassen. Gibt es eine Möglichkeit sich auch den Fehler (401, 500, ...) oder nähere Logs anzuschauen. Im NodeRed Log finde ich nichts, deshalb bin ich mir nicht mal sicher ob der Callback von der Jung Cloud über den MyFritz-DynDNS Eintrag überhaupt ankommt.

**Comment by Manu4771** [comment]
Hallo Crazy Bird, 

das Probelm mit der Aurhentifizierung konnte ich lösen. Das ist bei Node-RED ja eine sehr einfache HTTP-Authentifizierung nach RFC7616 / 7617: http(s):// + Username + Passwort + @ + Servername/

Du kannst PW und Benutzername also in Klartext in die URL legen. So ganz sicher finde ich das nicht. Aber wenn diue URL die richtige Syntax hat kommt man damit direkt auf die Node-RED installation. 

Danke für den Tipp mit den Subscriptions, da haben sich bei mir schon hunderte Fehler angehäuft. für zwei Datenpunkte. Das heißt für mich, dass aus dem Home Assitsent heraus tatsächlich was bei Jung ankommt. Aber was? 

Da ich Anfänger mit Node-RED bin und werde ich versuchen das Problem von der Seite anzugehen. Bei allem anderen bin ich jetzt erstmal am Ende mit meinem Latein. 

Ich finde super, das Jung OpenAPI bedient und sich damit für quasi alle Verwendungen öffnet. Wäre nur super ich würde es zum Laufen bekommen.

**Comment by patrickkaucher** [comment]
Hallo zusammen,

ich bin auch blutiger Anfänger mit Node Red.

Ich bekomme nichtmal meine Steckdosen geschalten. Das Feedback vom Debug sieht zwar gut aus, es passiert jedoch nichts :D 

Anbei der Auszug:

12/13/2024, 5:42:37 PM[node: Wohnzimmer Beamer Switch](http://192.168.2.86:8123/api/hassio_ingress/QV6j0PfE3KgfhO-VZnVLiRWHoRBrCW0c6jtjdrfJwl4/#)
msg : string[70]
"jung-sendCommand> sending PUT to=Wohnzimmer Beamer Switch, with cmd=on"

12/13/2024, 5:42:37 PM[node: Wohnzimmer Beamer Switch](http://192.168.2.86:8123/api/hassio_ingress/QV6j0PfE3KgfhO-VZnVLiRWHoRBrCW0c6jtjdrfJwl4/#)
msg : string[55]
"jung-sendCommand.after return> msg.statusCode=204 -> ok"

Vielleicht bin ich auch komplett falsch gewickelt.

**Comment by CrazyBird159** [comment]
Hallo Manu4771,
danke für den Hinweise mit dem Usernamen + Passwort in der URL. Damit komme ich jetzt auch an der Authentifizierung vorbei.

Mein Verständnis von den Subscriptions ist, dass Node Red die einmalig anlegt, wenn man die Node konfiguriert. Danach fragt Node Red aber nicht regelmäßig die Jung-Open-API an, sondern diese wird von sich aus aktiv, wenn sie vom Gerät die Info bekommt, dass sich z.B. die Temperatur oder der eingestellte Soll-Wert geändert hat. Dann versucht sie Subscription die Callback-URL aufzurufen und den neuen Wert zu übermitteln. Wenn etwas nicht passt, gibt es einen Fehler.

Auch mit Username + Passwort laufen in den Subscriptions weiter die Fehler hoch und ich bekomme keine Werte in Node Red.

Vielleicht ist mein Verständnis hier aber falsch. Ich bin aber auch Anfänger mit Note Red und mit APIs.

**Comment by CrazyBird159** [comment]
Ich hab die Lösung gefunden.

Als Callback URL  muss man http://[meinDNS oder meineIP]:[Port)/endpoint/ nutzen.
Dort benötigt man auch kein Username + Passwort.

**Comment by patrickkaucher** [comment]
Nice CrazyBird,

mit den Eckigen und den normalen Klammern?

**Comment by CrazyBird159** [comment]
Sollten alles eckige Klammern sein oben. Die sollten aber eigentlich nur anzeigen dass es ein Platzhalter ist.
 Also z.b. http://abcdefgzxywv.myfritz.net:1880/endpoint/

**Comment by patrickkaucher** [comment]
Irgendwie klappt das immer noch nicht bei mir. Habe alles ausprobiert. Alle möglichen Portweiterleitungen.
Wäre cool, wenn du kurz deine Konfiguration teils @Manu4771 

Sorry wollte eigentlich @CrazyBird159 verlinken

**Comment by Manu4771** [comment]
> Sollten alles eckige Klammern sein oben. Die sollten aber eigentlich nur anzeigen dass es ein Platzhalter ist. Also z.b. http://abcdefgzxywv.myfritz.net:1880/endpoint/

Bei mir klappt das leider auch nocht nicht. Was genau meinst du mit Endpoint? Die entgültige URL wird ja vom Programm selbst kreiert...

**Comment by patrickkaucher** [comment]
> > Sollten alles eckige Klammern sein oben. Die sollten aber eigentlich nur anzeigen dass es ein Platzhalter ist. Also z.b. http://abcdefgzxywv.myfritz.net:1880/endpoint/
> 
> Bei mir klappt das leider auch nocht nicht. Was genau meinst du mit Endpoint? Die entgültige URL wird ja vom Programm selbst kreiert...

Das /endpoint/ wird für anscheinend für den homeassistant benötigt.

**Comment by Manu4771** [comment]
okay. Der Endpoint war kein Platzhalter - verstanden. Ich bekomme denoch weiterhin die Msg von zu beginn.

**Comment by Manu4771** [comment]
Bei mir scheint es nun zu funktioniere. Die von mir fest vergebene URL zur Node-RED installatuion habe ich einfach mit /endpint/ ergänzt. 

In den Subscriptions bei der Jung API sehe ich nun die beiden Datenpunkte die ich testweise abgegriffen habe und jeweils mit 0 Fehler. Das scheint also zu funktionieren. Jetzt bräucht ich nur noch support wie ich dem Temp-Sensor und die Rolladenpostion in HA-Entitäten übersetze. aktuell bekomme ich noch keine Werte angezeigt.

**Comment by patrickkaucher** [comment]
> Bei mir scheint es nun zu funktioniere. Die von mir fest vergebene URL zur Node-RED installatuion habe ich einfach mit /endpint/ ergänzt. 
> 
> 
> 
> In den Subscriptions bei der Jung API sehe ich nun die beiden Datenpunkte die ich testweise abgegriffen habe und jeweils mit 0 Fehler. Das scheint also zu funktionieren. Jetzt bräucht ich nur noch support wie ich dem Temp-Sensor und die Rolladenpostion in HA-Entitäten übersetze. aktuell bekomme ich noch keine Werte angezeigt. 


Soweit bin ich auch schon, das bedeutet glaub nur das die Substribtion gesetzt wurde. Soweit bin ich jetzt auch, bei mir ist es aber eine Steckdose die geschalten wird. 

Betätige ich die Steckdose jetzt, kommt ein Fehler, den ich nicht nachvollziehen kann.

**Comment by patrickkaucher** [comment]
Ich hab's für mich jetzt erstmal gelöst. 


Callback URL http://dein-host/endpoint/

Dann spuckt die Receive Status node die endgültige URL aus für das Gerät.

Dann eine http in Node mit der URL (ohne den Host und ohne /endpoint)

Und da kommen die Daten an.

**Comment by patrickkaucher** [comment]
@Manu4771 hast du es hinbekommen?

**Comment by Tulkas1234** [comment]
> Ich hab's für mich jetzt erstmal gelöst.
> 
> Callback URL http://dein-host/endpoint/
> 
> Dann spuckt die Receive Status node die endgültige URL aus für das Gerät.
> 
> Dann eine http in Node mit der URL (ohne den Host und ohne /endpoint)
> 
> Und da kommen die Daten an.

Kann mir einer weiterhelfen wie ich das verschalten muss?

<img width="1264" height="813" alt="Image" src="https://github.com/user-attachments/assets/7f93da20-4154-495e-bba9-dd48fa7a506d" />

**Comment by patrickkaucher** [comment]
@Tulkas1234 hast du dein Problem gelöst?
So sieht die Konfiguration bei mir aus:
 
<img width="1251" height="833" alt="Image" src="https://github.com/user-attachments/assets/d6d422d3-65cf-4a7c-b3ec-b899d49c0fdf" />

**Comment by Tulkas1234** [comment]
Vielen Dank ich werde die Tage einen neuen Versuch starten, das System in Home Assistant über NodeRed zu integrieren. 

Aktuell habe ich alles über Smartthings eingebunden, aber eine direkte Verbindung über NodeRed ist sicherlich besser
